thread: add granular critical section to nonblocking collectives

hzhou opened the pull request:
## Background
There are two main components for thread safety: communication path and non-communication path. 

Non-communication path is non-performance critical, and it can benefit from coarse thread granularity -- simpler code as well as better *serial* performance due to fewer locks.

The communication path is performance-critical and incurring unnecessary serialization is detrimental. This path needs fine-grained critical sections -- typically tight lock-unlock blocks around netmod API calls.

There are 3 main communication paths: pt2pt, collective, and RMA.

## Description
This PR deals with collective path.
Add granular critical section to nonblocking collectives and HCOLL.

## Question for debate
* Why is this PR necessary?
  Collectives will eventually all decompose into pt2pt calls. All pt2pt calls will be made thread-safe and fine-grained. 
** If collectives take on global lock, it will not block other pt2pt calls because pt2pt calls does not use global lock. It will serialize collectives and other MPI operations such as datatype and communicator functions.
** If collective take on "collective" lock, the previous point largely applies plus collective call can happen in parallel with other threads doing datatype and communicator calls.
** Collective calls cannot overlap per communicator, but can happen in parallel on different communicators.
## TODO
* [ ] Remove global locks in nonblocking collectives
* [ ] Remove global locks in blocking collectives

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Passes tests (included warning check)
* [ ] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [ ] Commits are self-contained and do not do two things at once
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

hzhou commented:
test:mpich/warnings

hzhou commented:
test:mpich/warnings

hzhou commented:
@rzambre These are all global locks that in the sense that we need them regardless `MPICH_THREAD_CS_GRANULARITY__VCI` or `MPICH_THREAD_CS_GRANULARITY__GLOBAL`, right? Should we use `MPID_THREAD_CS_ENTER(GLOBAL, ...)` then?

rzambre commented:
These locks are not needed in the `GLOBAL` critical section since access to the MPI library itself is serialized. For the `per-VCI` critical section, we could have multiple threads issuing non-blocking collectives on different communicators on different transports (parallel `Sched` or `HCOLL` non-blocking collectives).

Also, we would need another commit that removes the use of the `GLOBAL` lock for non-blocking collective operation in the `per-VCI` critical section model. If there are no conflicts, you could cherry-pick this one: https://github.com/pmodels/mpich/pull/3659/commits/94583000a75668d93e375bc0715d39da30890b3d

rzambre commented on src/mpid/common/sched/mpidu_sched.c:
```diff
@@ -136,13 +136,19 @@ struct MPIDU_Sched_state {
 
 /* holds on to all incomplete schedules on which progress should be made */
-struct MPIDU_Sched_state all_schedules = { NULL };
+static struct MPIDU_Sched_state all_schedules = { NULL };
```
I think we should add a namespace to `all_schedules` just like we did for `coll_queue`.

hzhou replied:
We don't need namespace for static variables.

rzambre replied:
Got it

rzambre commented on src/mpid/ch4/src/ch4_types.h:
```diff
@@ -322,12 +322,17 @@ extern MPL_dbg_class MPIDI_CH4_DBG_MAP;
 extern MPL_dbg_class MPIDI_CH4_DBG_COMM;
 extern MPL_dbg_class MPIDI_CH4_DBG_MEMORY;
 #endif
+
```
Extra line crept in.

hzhou replied:
resolved with new commit

rzambre commented on src/mpid/ch4/src/ch4_types.h:
```diff
@@ -328,5 +328,6 @@ extern MPL_dbg_class MPIDI_CH4_DBG_MEMORY;
 /* Protects MPIDIG global structures (e.g. global unexpected message queue) */
 #define MPIDIU_THREAD_MPIDIG_GLOBAL_MUTEX  MPIDI_global.m[3]
 #define MPIDIU_THREAD_SCHED_LIST_MUTEX  MPIDI_global.m[4]
+#define MPIDIU_THREAD_TSP_QUEUE_MUTEX  MPIDI_global.m[3]
```
Oops, should be `MPIDI_global.m[5]`.

hzhou replied:
resolved with new commit

rzambre commented on src/mpid/ch4/src/ch4_types.h:
```diff
@@ -322,12 +322,17 @@ extern MPL_dbg_class MPIDI_CH4_DBG_MAP;
 #define MPIDIU_THREAD_MPIDIG_GLOBAL_MUTEX  MPIDI_global.m[3]
 #define MPIDIU_THREAD_SCHED_LIST_MUTEX  MPIDI_global.m[4]
-#define MPIDIU_THREAD_TSP_QUEUE_MUTEX  MPIDI_global.m[3]
+#define MPIDIU_THREAD_TSP_QUEUE_MUTEX  MPIDI_global.m[5]
```
With the fix in the previous commit, the diff on this line would go away.

hzhou replied:
resolved with new commit

rzambre commented on src/mpid/common/hcoll/hcoll_init.c:
```diff
@@ -48,7 +48,6 @@ MPL_dbg_class MPIR_DBG_HCOLL;
 
 void hcoll_rte_fns_setup(void);
 
-
```
Line deleted.

raffenet commented:
@sagarth please take a look (and accept your mpich-intel team invitation so we can assign reviews 😄 )

rzambre commented on src/mpid/common/sched/mpidu_sched.c:
```diff
@@ -488,10 +494,12 @@ int MPIDU_Sched_start(MPIR_Sched_t * sp, MPIR_Comm * comm, int tag, MPIR_Request
     DL_APPEND(all_schedules.head, s);
+    MPID_THREAD_CS_EXIT(VCI, MPIDIU_THREAD_SCHED_LIST_MUTEX);
+
+    MPID_Progress_activate_hook(MPIR_Nbc_progress_hook_id);
```
Could we add a comment about why the check for `all_schedules.head == NULL` is removed? Would be useful for future readers.

rzambre replied:
Actually, that explanation should go in the commit message.

sagarth replied:
Also the order of calling `DL_Append` and `MPID_Progress_activate_hook` are changed. Were these changes  result of adding the critical section?

hzhou replied:
Good that you spotted it :) Originally it has a check for `all_schedules.head==NULL` and only activate it when it is NULL. I removed that check since I reason that inside `MPID_Progress_activate_hook` will do a check anyway, so the check doesn't seem to save anything -- assuming the atomic load inside `MPID_Progress_activate_hook` is really cheap anyway. 

I swapped the order on the reasoning that between the activating hook and appending schedules, another thread may deactivate the hook in-between.

Does that make sense?

sagarth replied:
I see. Thanks for explaining. Do you think we can treat these changes as orthogonal to the current patch (adding granular CS..) and move to a separate commit?

rzambre commented on src/mpi/coll/transports/gentran/tsp_gentran.c:
```diff
@@ -429,9 +429,12 @@ int MPII_Genutil_sched_start(MPII_Genutil_sched_t * sched, MPIR_Comm * comm, MPI
     DL_APPEND(coll_queue.head, &(reqp->u.nbc.coll));
+    MPID_THREAD_CS_EXIT(VCI, MPIDIU_THREAD_TSP_QUEUE_MUTEX);
+
+    MPID_Progress_activate_hook(MPII_Genutil_progress_hook_id);
```
Same here. A comment about why we removed the check for `coll_queue.head == NULL` in the commit message would be helpful.

rzambre commented on src/mpid/common/hcoll/hcoll_init.c:
```diff
@@ -158,6 +157,7 @@ int hcoll_comm_create(MPIR_Comm * comm_ptr, void *param)
         comm_ptr->hcoll_priv.is_hcoll_init = 0;
         goto fn_exit;
     }
+
```
Extra line crept in.

raffenet commented:
@sagarth yes, to whichever email is associated with your Github account. There is a probably a notification waiting for you somewhere in your account settings/memberships.

sagarth commented:
@raffenet Hmm... I could not find it in the mailbox of the associated email or under the github settings. Is it possible for you to send the invite again?

hzhou commented:
test:jenkins/ch3/sock
test:jenkins/ch4/ofi

hzhou commented:
@rzambre I modified and added some notes to the description. I think it remains to be discussed why this PR is necessary. Collectives cannot overlap due to semantics by MPI standards, so a coarser collective or even global lock seems desirable. Collectives will eventually reduce to pt2pt operations, which does not (we'll ensure that) interfere with global locks, so it should work (in principle).

rzambre commented:
The standard doesn't allow for parallel collectives on the same communicator but we can have parallel collectives on different communicators.

Collectives boil down to pt2pt operations. So, thread-safety during the issue of collective operations would be maintained by the thread-safety of pt2pt operations.

Most of this PR is for the thread-safety during the progress of collectives, which is required.

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

hzhou commented:
test:jenkins/ch3/sock

raffenet commented:
> The standard doesn't allow for parallel collectives on the same communicator but we can have parallel collectives on different communicators.

The standard says collectives must be issued in the same order for all processes in a communicator. You can certainly have multiple outstanding collectives progressing on a communicator. Just want to make sure that's clear.

rzambre commented:
Correct, there can multiple outstanding collectives on a communicator on a process. But the standard doesn't allow the parallel issue of collectives on the same communicator, correct?

raffenet commented:
Collective issue must be serialized on a communicator, yes. That is the user's responsibility.

hzhou commented:
test:jenkins/ch4/ofi

hzhou commented:
test:jenkins/ch3/most

hzhou commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

rzambre commented:
Looks good to me!
